### PR TITLE
qemu_img_check_data_integrity: fix leaked clusters issue

### PR DESCRIPTION
**Pull request by hellohellenmao**
1. Backup image before testing and restore after
2. Check the image and repair the leaked clusters after
data integrity testing

Signed-off-by: Tingting Mao <timao@redhat.com>

ID: 2103037

### CODE REVIEW

**Comment by hellohellenmao** [comment]
Depends on avocado-framework/avocado-vt/pull/3453

**Comment by hellohellenmao** [comment]
(1/2) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.qemu_img_check_data_integrity.kill_qemu_during_io.q35: STARTED
 (1/2) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.qemu_img_check_data_integrity.kill_qemu_during_io.q35: PASS (360.65 s)
 (2/2) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.qemu_img_check_data_integrity.kill_qemu_after_io.q35: STARTED
 (2/2) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.qemu_img_check_data_integrity.kill_qemu_after_io.q35: PASS (365.88 s)

**Comment by hellohellenmao** [comment]
(1/2) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.qemu_img_check_data_integrity.kill_qemu_during_io.q35: STARTED
 (1/2) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.qemu_img_check_data_integrity.kill_qemu_during_io.q35: PASS (370.76 s)
 (2/2) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.qemu_img_check_data_integrity.kill_qemu_after_io.q35: STARTED
 (2/2) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.qemu_img_check_data_integrity.kill_qemu_after_io.q35: PASS (372.50 s)

File: qemu/tests/qemu_img_check_data_integrity.py [hunk 1]
```diff
@@ -89,3 +92,17 @@ def run_iozone_background(vm):
         session = vm.wait_for_login()
         session.cmd(params["rm_testfile_cmd"] % iozone_testfile)
     session.close()
+    vm.destroy()
```
Comment (zhencliu) [inline qemu/tests/qemu_img_check_data_integrity.py:95]: hi @hellohellenmao , if we don't quit vm gracefully(e.g. gracefully=False), is that much more likely to hit the 'leaked clusters'?

File: qemu/tests/qemu_img_check_data_integrity.py [hunk 2]
```diff
@@ -89,3 +92,17 @@ def run_iozone_background(vm):
...
+    img_param = params.object_params("image1")
+    img = QemuImg(img_param, root_dir, "image1")
+    res = img.check(params, root_dir)
+    if res.exit_status == 2:
+        test.fail("Image is corrupted, check please.")
+    elif res.exit_status == 3:
+        test.log.info("Leaked clusters found in the image, repair them now.")
+        res = img.check(params, root_dir, check_repair="leaks")
+        if res.exit_status != 0:
+            test.fail("Repair the leaked image failed, check please.")
```
Comment (zhencliu) [inline qemu/tests/qemu_img_check_data_integrity.py:108]: do we need additional check besides the image.check? e.g. 
    start vm again and try to log into it, there should no error
That can make sure repairing the image doesn't lead to a corruption.

File: qemu/tests/qemu_img_check_data_integrity.py [hunk 3]
```diff
@@ -19,6 +21,7 @@ def run(test, params, env):
     5. If tmp_file_check == no
     5.1 Kill qemu process during writing data in the guest
     6. Boot the guest again, make sure it could boot successfully
+    7. Check the testing image and repair the leaked clusters
```
Comment (XueqiangWei) [inline qemu/tests/qemu_img_check_data_integrity.py:24]: A suggestion, add steps "backup image before testing" and "restore image after testing" into Doc string.

File: qemu/tests/qemu_img_check_data_integrity.py (hunk 2 above)
Comment (XueqiangWei) [reply qemu/tests/qemu_img_check_data_integrity.py:108]: same question as zhenchao, do we need to start vm again?

File: qemu/tests/qemu_img_check_data_integrity.py (hunk 1 above)
Comment (hellohellenmao) [reply qemu/tests/qemu_img_check_data_integrity.py:95]: Yeah, it should be. But I do not think we need it. The leaked clusteres should be created while `kill_vm_process` in previous steps, but not in this step.

File: qemu/tests/qemu_img_check_data_integrity.py (hunk 2 above)
Comment (hellohellenmao) [reply qemu/tests/qemu_img_check_data_integrity.py:111]: Updated, thanks.

**Comment by hellohellenmao** [comment]
(1/2) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.qemu_img_check_data_integrity.kill_qemu_during_io.q35: STARTED
 (1/2) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.qemu_img_check_data_integrity.kill_qemu_during_io.q35: PASS (538.21 s)
 (2/2) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.qemu_img_check_data_integrity.kill_qemu_after_io.q35: STARTED
 (2/2) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.qemu_img_check_data_integrity.kill_qemu_after_io.q35: PASS (392.45 s)

**Comment by hellohellenmao** [comment]
@zhencliu @XueqiangWei Could you please help to review again? Thanks.

**Comment by zhencliu** [comment]
LGTM

**Comment by XueqiangWei** [comment]
LGTM,
Acked-by: xuwei@redhat.com

**Pull request merged**